### PR TITLE
Add edgex-secrets-token provider tests for consumer snaps

### PR DESCRIPTION
**Pull request by MonicaisHer**

### CODE REVIEW

File: test/suites/device-mqtt/main_test.go [hunk 1]
```diff
@@ -39,13 +43,6 @@ func TestMain(m *testing.M) {
 	// make sure all services are online before starting the tests
 	utils.WaitPlatformOnline(nil)
 
-	// for local build, the interface isn't auto-connected.
-	// connect manually regardless
-	utils.SnapConnect(nil,
-		"edgexfoundry:edgex-secretstore-token",
-		deviceMqttSnap+":edgex-secretstore-token",
-	)
```
Comment (farshidtz) [inline test/suites/device-mqtt/main_test.go:47]: The connection should happen here, because the tests depend on it. See other comments.

File: test/suites/device-mqtt/main_test.go [hunk 2]
```diff
@@ -61,6 +58,10 @@ func TestMain(m *testing.M) {
 }
 
 func TestCommon(t *testing.T) {
+	utils.TestSecret(t, deviceMqttApp, utils.Secret{
+		TestManualConnection: needManualConnection,
+	})
+
```
Comment (farshidtz) [inline test/suites/device-mqtt/main_test.go:64]: Does TestSecret run before all other tests? If not, the tests will fail when the snap is built locally.

File: test/utils/secret.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+				"edgex-"+consumerName+":edgex-secretstore-token",
+			)
+		}
+		requireCorrectSecret(t, consumerName)
+	})
+}
+
+func requireCorrectSecret(t *testing.T, consumerName string) {
+	providerSecret, stderr := Exec(t, "sudo cat /var/snap/edgexfoundry/current/secrets/"+consumerName+"/secrets-token.json")
+	require.Empty(t, stderr)
```
Comment (farshidtz) [inline test/utils/secret.go:26]: the Exec will raise an error if it can't ready. Checking stderr shouldn't be necessary.

File: test/utils/secret.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+func TestSecret(t *testing.T, consumerName string, conf Secret) {
+	t.Run("secrets interface", func(t *testing.T) {
+		if conf.TestManualConnection {
+			SnapConnect(nil,
+				"edgexfoundry:edgex-secretstore-token",
+				"edgex-"+consumerName+":edgex-secretstore-token",
+			)
+		}
+		requireCorrectSecret(t, consumerName)
```
Comment (farshidtz) [inline test/utils/secret.go:20]: I suggest moving this to main, after installing the snaps.

File: test/utils/secret.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	})
+}
+
+func requireCorrectSecret(t *testing.T, consumerName string) {
```
Comment (farshidtz) [inline test/utils/secret.go:24]:
```suggestion
func TestSeededSecretstoreToken(t *testing.T, consumerName string) {
```

**Review by farshidtz — CHANGES_REQUESTED** [review]

File: test/utils/secret.go (hunk 3 above)
Comment (farshidtz) [reply test/utils/secret.go:26]: Maybe add a check to make sure providedSecret isn't empty. Otherwise, an empty file copied over will pass the tests.

File: test/suites/device-mqtt/main_test.go (hunk 2 above)
Comment (MonicaisHer) [reply test/suites/device-mqtt/main_test.go:64]: Yes, TestSecret runs before all other tests.

File: test/utils/secret.go (hunk 4 above)
Comment (MonicaisHer) [reply test/utils/secret.go:20]: I have moved it to main, under the SnapInstallFromFile section, because the Snap connection is only needed when installing from a file. Does this sound good?

File: test/utils/secret.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package utils
+
+import (
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestSecret(t *testing.T, consumerName string) {
```
Comment (farshidtz) [inline test/utils/secret.go:8]: We may wanna name this in a more generic way, such as TestContentInterfaces to allow having both secret seeding and config interfaces inside. 

We can do it once we have tests for the config interface.

**Review by farshidtz — APPROVED** [review]
Looks good. Please see comments.

File: test/utils/secret.go (hunk 4 above)
Comment (farshidtz) [reply test/utils/secret.go]: Yes. Looks cleaner.

**Comment by MonicaisHer** [comment]
@farshidtz Thanks for your re-review. Please don't merge it yet, because I would duplicate the secret test for the rest test suites in this PR.

File: test/utils/secret.go (hunk 6 above)
Comment (MonicaisHer) [reply test/utils/secret.go:8]: make sense, I will add this as a TODO comment in the test.

File: test/utils/secret.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+	"testing"
+)
+
+type Secret struct {
+	TestSecretsInterface bool
+}
+
+// TODO: rename to a more generic name such as TestContentInterfaces
+// to allow having both secret seeding and config interfaces inside
+func TestSecret(t *testing.T, providerPath, consumerName, consumerSecretPath string, conf Secret) {
```
Comment (farshidtz) [inline test/utils/secret.go:14]: Why some of the configurations are in the config object, some others as function arguments?

File: test/utils/secret.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package utils
+
+import (
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+type Secret struct {
+	TestSecretsInterface bool
```
Comment (farshidtz) [inline test/utils/secret.go:9]: How about 
```suggestion
	TestSecretToken string // path to where the token is expected
```

File: test/suites/device-virtual/main_test.go [hunk 9]
```diff
@@ -38,19 +45,6 @@ func TestMain(m *testing.M) {
...
-		"edgexfoundry:edgex-secretstore-token",
-		deviceVirtualSnap+":edgex-secretstore-token",
-	)
-
-	// Start the service so that the default config gets uploaded to consul.
-	// Otherwise, settings that get passed using environment variables on first start get uploaded
-	// and become the default.
-	utils.SnapStart(nil, deviceVirtualService)
-	utils.WaitServiceOnline(nil, 60, deviceVirtualServicePort)
-
```
Comment (farshidtz) [inline test/suites/device-virtual/main_test.go:53]: Why are these removed?

File: test/utils/secret.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+		} else {
+			t.Skip("Security off mode, there is no secret needed.")
+		}
+	})
+}
+
+func TestSeededSecretstoreToken(t *testing.T, providerPath, consumerName, consumerPath string) {
+	t.Run("same content on both sides", func(t *testing.T) {
+		providerSecret, _ := Exec(t, "sudo cat /var/snap/edgexfoundry/current/secrets/"+providerPath+"/secrets-token.json")
+		consumerSecret, _ := Exec(t, "sudo cat /var/snap/"+consumerName+"/current/"+consumerPath+"/secrets-token.json")
```
Comment (farshidtz) [inline test/utils/secret.go:27]: Confusing use of variable names. These aren't paths but rather just identical app name.

File: test/suites/device-virtual/main_test.go (hunk 9 above)
Comment (MonicaisHer) [reply test/suites/device-virtual/main_test.go:53]: They have been handled by the beginning of config test now:
https://github.com/canonical/edgex-snap-testing/blob/dbc11de74b039af37f0d5731855e62659739c759/test/utils/config.go#L29-L33

File: test/utils/secret.go (hunk 10 above)
Comment (MonicaisHer) [reply test/utils/secret.go:27]: How about:
```
type Secret struct {
	TestSecretsInterface bool
	ConsumerName         string // e.g. edgex-app-service-configurable, edgex-ekuiper, edgex-device-mqtt
	ConsumerAppName      string // e.g. app-rules-engine,               edgex-ekuiper, device-mqtt
}
```

File: test/suites/device-virtual/main_test.go (hunk 9 above)
Comment (farshidtz) [reply test/suites/device-virtual/main_test.go:53]: great. thanks

File: test/utils/secret.go (hunk 10 above)
Comment (farshidtz) [reply test/utils/secret.go:27]: I think we can go with Snap and App instead of ConsumerName and ConsumerAppName.

There is no plural "secrets" interface. We have a content interface named edgex-secretstore-token. So maybe: TestSecretsInterface -> TestSecretToken

**Comment by farshidtz** [comment]
I did some refactoring to allow extensions. 

I removed `t.Skip("Security off mode, there is no secret needed.")` because setting enabling/disabling the test does not always imply that security is on/off.

**Review by farshidtz — COMMENTED** [review]
Looks good.
